src/Features/Core/Portable/Completion/Providers/ImportCompletionProvider/ExtensionMethodImportCompletionHelper.cs

```diff
@@ -67,190 +68,187 @@ async Task<(ImmutableArray<SerializableImportCompletionItem>, StatisticCounter)>
             counter.TotalExtensionMethodsProvided = items.Length;
             counter.Report();
 
-            return items;
+            return (items, counter);
         }
 
-        /// <summary>
-        /// Get the metadata name of all the base types and interfaces this type derived from.
-        /// </summary>
-        private static ImmutableArray<string> GetReceiverTypeNames(ITypeSymbol receiverTypeSymbol)
+        public static async Task<(ImmutableArray<SerializableImportCompletionItem>, StatisticCounter)> GetUnimportedExtensionMethodsInCurrentProcessAsync(
+            Document document,
+            int position,
+            ITypeSymbol receiverTypeSymbol,
+            ISet<string> namespaceInScope,
+            bool forceIndexCreation,
+            CancellationToken cancellationToken)
         {
-            using var _ = PooledHashSet<string>.GetInstance(out var allTypeNamesBuilder);
-            AddNamesForTypeWorker(receiverTypeSymbol, allTypeNamesBuilder);
-            return allTypeNamesBuilder.ToImmutableArray();
+            var counter = new StatisticCounter();
+            var ticks = Environment.TickCount;
 
-            static void AddNamesForTypeWorker(ITypeSymbol receiverTypeSymbol, PooledHashSet<string> builder)
+            // First find symbols of all applicable extension methods.
+            // Workspace's syntax/symbol index is used to avoid iterating every method symbols in the solution.
+            var results = await GetExtensionMethodSymbolsAsync(
+                document,
+                position,
+                receiverTypeSymbol,
+                namespaceInScope,
+                forceIndexCreation,
+                cancellationToken).ConfigureAwait(false);
+
+            counter.GetSymbolsTicks = Environment.TickCount - ticks;
+            ticks = Environment.TickCount;
+
+            var indicesComplete = true;
+            using var _1 = PooledDictionary<INamespaceSymbol, string>.GetInstance(out var namespaceNameCache);
+            using var _2 = PooledDictionary<(string containingNamespace, string methodName, bool isGeneric), (IMethodSymbol bestSymbol, int overloadCount)>.GetInstance(out var overloadMap);
```

Comment:
seriously on the fence about asking for some named structs.

Reply:
ValueTuple make it easier to use them as dictionary key

Reply:
understood.  still on the fence :)

Reply:
When will we start dogfooding records? :)